docs/cudf/source/developer_guide/pylibcudf.md

```diff
@@ -0,0 +1,113 @@
+In addition to these requirements, pylibcudf must also integrate naturally with other Python libraries.
+In other words, it should interoperate fairly transparently with standard Python containers, community protocols like `__cuda_array_interface__`, and common vocabulary types like CuPy arrays.
+
+
+## General Design Principles
+
+To satisfy the goals of pylibcudf, we impose the following set of design principles:
+- Every public function or method should be `cpdef`ed. This allows it to be used in both Cython and Python code. This incurs some slight overhead over `cdef` functions, but we assume that this is acceptable because 1) the vast majority of users will be using pure Python rather than Cython, and 2) the overhead of a `cpdef` function over a `cdef` function is on the order of a nanosecond, while CUDA kernel launch overhead is on the order of a microsecond, so these function overheads should be washed out by typical usage of pylibcudf.
+- Every variable used should be strongly typed and either be a primitive type (int, float, etc) or a cdef class. Any enums in C++ should be mirrored using `cpdef enum`, which will create both a C-style enum in Cython and a PEP 435-style Python enum that will automatically be used in Python.
+- All typing in code should be written using Cython syntax, not PEP 484 Python typing syntax. Not only does this ensure compatibility with Cython < 3, but even with Cython 3 PEP 484 support remains incomplete as of this writing.
+- All cudf code should interact only with pylibcudf, never with libcudf directly.
+- All imports should be relative so that pylibcudf can be easily extracted from cudf later
+  - Exception: All imports of libcudf API bindings in `cudf._lib.cpp` should use absolute imports of `cudf._lib.cpp as libcudf`. We should convert the `cpp` directory into a proper package so that it can be imported as `libcudf` in that fashion. When moving pylibcudf into a separate package, it will be renamed to `libcudf` and only the imports will need to change.
+- Ideally, pylibcudf should depend on nothing other than rmm and pyarrow. This will allow it to be extracted into a a largely standalone library and used in environments where the larger dependency tree of cudf may be cumbersome.
+
+
+## Relationship to libcudf
+
+In general, the relationship between pylibcudf and libcudf can be understood in terms of two components, data structures and algorithms.
+
+(data-structures)=
+
+### Data Structures
+
+Typically, every type in libcudf should have a mirror Cython `cdef` class with an attribute `self.c_obj: unique_ptr[${underlying_type}]` that owns an instance of the underlying libcudf type.
+Each type should also implement a corresponding method `cdef ${cython_type} from_${underlying_type}(${underlying_type} dt)` to enable constructing the Cython object from ann underlying libcudf instance.
```

Comment:
Are these now always just called `from_libcudf`? The repetition of the type name in the method seems somewhat redundant:
```cython
cdef class Column:
    @staticmethod
    cdef Column from_column(cudf::column dt)
     ....
   @staticmethod
    cdef Column from_libcudf(cudf::column dt)

...
col = Column.from_libcudf(..)
col = Column.from_column(...)
```
?

Reply:
Yup we want `from_libcudf` now, will update.